lua/pac3/core/client/parts/animation.lua

```diff
@@ -2,6 +2,79 @@ local FrameTime = FrameTime
 
 local BUILDER, PART = pac.PartTemplate("base")
 
+local AnimStack = {
+	__index = {
+		push = function(self, part)
+			local stack = self.stack
+
+			if #stack == 0 then
+				-- Empty stack
+				table.insert(stack, part)
+			else
+				-- Stop the current animation if it's not self
+				local top = stack[#stack]
+			
+				-- Remove invalid parts
+				while top and not top:IsValid() do
+					top = table.remove(stack)
+				end
+			
+				if top ~= part then
+					if top then
+						top:OnStackStop()
+						top.pac_animation_stack_current = false
+					end
+				
+					if part.pac_animation_stack_contains then
+						-- Check this variable to save some perf
+						-- Remove self from stack to move to end and also prevent things from breaking because table.RemoveByValue() only removes the first instance
+						table.RemoveByValue(stack, part)
+					end
+
+					table.insert(stack, part)
+				end
+			end
+
+			part:OnStackStart()
+			part.pac_animation_stack_current = true
+			part.pac_animation_stack_contains = true
```

Comment:
One last thing actually. These can be members of this stack object instead of members of the part. Mind doing that? It would also be good if there was `getTop = function(self)` so you could do `local top = self:getTop()`

Reply:
How could these be part of the stack when they need to be unique for each part and the stack is only unique for the part's owner?

Reply:
And also if I move them into AnimStack.stack that defeats the entire point of the flags which is to avoid having to search the whole table when the part is potentially not in it

Reply:
You could set `self.current_part = part` and replace stack_contains with `self.parts[part]`

Reply:
The stack_contains optimization could be removed anyway though. It's not like we're expecting 1000 things on the stack